File: 1338-reduce-array-size-to-the-half/1338-reduce-array-size-to-the-half.js
/**
 * @param {number[]} arr
 * @return {number}
 */
function counter(arr) {
    let hashs = {};
    for (num of arr){
        if (num in hashs) {
            hashs[num] += 1;
        } else {
            hashs[num] = 1;
        }
    }
    return hashs;
};

var minSetSize = function(arr) {
    let hashs = counter(arr);
    let keys = Object.keys(hashs);
    keys.sort((curr_key, other_key) => (hashs[other_key] - hashs[curr_key]))
    let goal = (arr.length/2).toFixed(0);
    let result = 0;
    for (let key of keys){
        result += 1;
        goal -= hashs[key];
        if (goal <= 0) {
            break;
        }
    }
    return result;
};

File: 1338-reduce-array-size-to-the-half/1338-reduce-array-size-to-the-half.py
from collections import Counter
class Solution:
    def minSetSize(self, arr: List[int]) -> int:
        result = 0
        goal = len(arr) // 2
        print(goal)
        counter = Counter(arr)
        for num, count in counter.most_common():
            result += 1
            goal -= count
            if goal <= 0:
                break
            
        return result